### PR TITLE
nextMillennium Bid Adapter: Add usersync URL

### PR DESCRIPTION
**Pull request by mifanich**

### CODE REVIEW

**Comment by SyntaxNode** [comment]
I tested this new user sync url. I am able to see the expected 302 redirect, but I don't see a user id. You removed the user macro. Without a user macro, it is expected that the user id be appended to the end of the redirect url and i'm not seeing that behavior.

**Comment by SyntaxNode** [comment]
This is not a user sync url. A user sync url must redirect back to the Prebid Server host with a single user id for the adapter which will be stored in the federated Prebid Server cookie.

This is a reference to a forked `load-cookie.html`. I do not believe it is appropriate to use the cookie sync mechanism in this way, as it will cause double the amount of cookie syncs. Perhaps you can work with with your publishers to add a reference to their page as is done for `load-cookie.html` today or you can talk with the Universal Creative repo folks about including your PBS host in that list.

**Comment by hhhjort** [comment]
> > 
> 
> @SyntaxNode Thank you for reply! it is important for our service that prebid could initiate cookie_sync. Results are generated by load-cookie.html could help for correct work of our PBS.

I think you are a bit confused here. Yes, load-cookie.html is very useful for publishers to implement to start the cookie sync process. load-cookie.html will call the cookiesync endpoint on PBS, which will then return to the caller the user sync pixels needed to fill in the missing user syncs. PBS uses the `userSync` section of your yaml file to find your usersync pixel. So configuring load-cookie.html here will tell load-cookie.html to call load-cookie.html again. This would result in an infinite loop. What you should put in `userSync` is the pixel that will sync your userID to the PBS user, so that PBS can give you the correct `buyerID` in the request it sends you.

**Comment by mifanich** [comment]
> > > 
> > 
> > 
> > @SyntaxNode Thank you for reply! it is important for our service that prebid could initiate cookie_sync. Results are generated by load-cookie.html could help for correct work of our PBS.
> 
> I think you are a bit confused here. Yes, load-cookie.html is very useful for publishers to implement to start the cookie sync process. load-cookie.html will call the cookiesync endpoint on PBS, which will then return to the caller the user sync pixels needed to fill in the missing user syncs. PBS uses the `userSync` section of your yaml file to find your usersync pixel. So configuring load-cookie.html here will tell load-cookie.html to call load-cookie.html again. This would result in an infinite loop. What you should put in `userSync` is the pixel that will sync your userID to the PBS user, so that PBS can give you the correct `buyerID` in the request it sends you.

@hhhjort Thanks a lot for the answer!
Infinite loop will not happened because our load-cookie.html makes redirects to our PBS, not yours. 
If you open our load-cookie.html, you can see it. Our flow suspects cookie_sync execution every auction.

**Comment by SyntaxNode** [comment]
> Infinite loop will not happen

I agree on this point. I do not see the potential for an infinite loop.

> Our flow suspects cookie_sync execution every auction.

The cookie syncing process is separate from the auction. Publishers typically run a cookie sync once per page and only a subset of adapters are chosen per run. The fact that the Next Millennium cookie sync process does not call back into the PBS setuid endpoint gives it an unfair advantage in that it will be chosen more often because PBS (whoever is hosting it) will never store a Next Millennium user id.

Beyond the fairness issue, I really don't see how you expect the load-cookie.html approach to work with your adapter. Your load-cookie.html script will populate a cookie under your PBS hosted domain. When a publisher sends a bid request to a different PBS host and calls your adapter, that request will not have your cookie. Even if we did allow this script for your cookie syncing endpoint, I can't see how it would be of benefit to your adapter.

Could you please explain end-to-end what you are trying to accomplish? Perhaps we can advise you on an alternate approach.

**Comment by mifanich** [comment]
> > Infinite loop will not happen
> 
> I agree on this point. I do not see the potential for an infinite loop.
> 
> > Our flow suspects cookie_sync execution every auction.
> 
> The cookie syncing process is separate from the auction. Publishers typically run a cookie sync once per page and only a subset of adapters are chosen per run. The fact that the Next Millennium cookie sync process does not call back into the PBS setuid endpoint gives it an unfair advantage in that it will be chosen more often because PBS (whoever is hosting it) will never store a Next Millennium user id.
> 
> Beyond the fairness issue, I really don't see how you expect the load-cookie.html approach to work with your adapter. Your load-cookie.html script will populate a cookie under your PBS hosted domain. When a publisher sends a bid request to a different PBS host and calls your adapter, that request will not have your cookie. Even if we did allow this script for your cookie syncing endpoint, I can't see how it would be of benefit to your adapter.
> 
> Could you please explain end-to-end what you are trying to accomplish? Perhaps we can advise you on an alternate approach.

@SyntaxNode 
Sorry for long reply. I have researched all you told me. So we have implemented a separate service, that returns th nextMIlennium UID according to the documentation

**Comment by SyntaxNode** [comment]
> @SyntaxNode Sorry for long reply. I have researched all you told me. So we have implemented a separate service, that returns the nextMIlennium UID according to the documentation

That looks much better. Thank you. I tested with the following URL which could be constructed by PBS :

```
https://cookies.nextmillmedia.com/sync?gdpr=&gdpr_consent=&us_privacy=&redirect=https%3A%2F%2Fib.adnxs.com%2Fprebid%2Fsetuid%3Fbidder%3Dnextmillennium%26gdpr%3D%26gdpr_consent%3D%26f%3Di%26uid%3D%5BNMUID%5D%0A
```

I receive an error from the browser regarding the SSL certificate and get a 404 from the endpoint.

File: static/bidder-info/nextmillennium.yaml [hunk 1]
```diff
@@ -7,3 +7,7 @@ capabilities:
   site:
     mediaTypes:
       - banner
+userSync:
+  iframe:
+    url: "https://cookies.nextmillmedia.com/sync?gdpr={{.GDPR}}&gdpr_consent={{.GDPRConsent}}&us_privacy={{.USPrivacy}}&redirect={{.RedirectURL}}"
```
[the remark this answers is not in the document]
Comment (mifanich) [reply static/bidder-info/nextmillennium.yaml:12]: > I tried testing this usersync URL and I am getting back a 404 response code

@mansinahar @SyntaxNode 
I have fixed 404 response code. You are welcome to continue testing

[the remark this answers is not in the document]
Comment (SyntaxNode) [reply static/bidder-info/nextmillennium.yaml:12]: Retested and looks good now.

**Review by SyntaxNode — APPROVED** [review]

**Review by mansinahar — APPROVED** [review]

**Pull request merged**